encoding/int_encode.go

```diff
@@ -1,30 +1,188 @@
-func (enc *IntEnc) Encode(writer io.Writer, values ForwardIntIterator) error {
-	return errors.New("not implemented")
+func (enc *IntEnc) Encode(
+	writer io.Writer,
+	valuesIt RewindableIntIterator,
+) error {
+	// Determine whether we want to do table compression
+	var (
+		max  int
+		curr int
+		idx  int
+	)
+	min := math.MaxInt64
+	dict := make(map[int]int)
+	for valuesIt.Next() {
+		curr = valuesIt.Current()
+		// Only grow the dict map if we are below dictEncodingMaxCardinalityInt.
+		// This way we track if we've exceeded the max # of uniques.
+		if len(dict) <= dictEncodingMaxCardinalityInt {
+			if _, ok := dict[curr]; !ok {
+				dict[curr] = idx
+				idx++
+			}
+		}
+
+		// Track max/min.
+		max = int(math.Max(float64(curr), float64(max)))
+		min = int(math.Min(float64(curr), float64(min)))
+	}
+
+	if valuesIt.Err() != nil {
+		return valuesIt.Err()
+	}
+
+	// Rewind iteration.
+	valuesIt.Rewind()
+
+	switch {
+	case len(dict) <= dictEncodingMaxCardinalityInt:
+		// Table encode if there is less than 256 unique valuesIt.
+		enc.metaProto.Encoding = encodingpb.EncodingType_DICTIONARY
+		// Min number of bits to encode each actual value into the table.
+		enc.metaProto.BytesPerDictionaryValue = int64(math.Ceil(float64(bits.Len(uint(max))) / 8.0))
+		// Min number of bits required to encode table indices.
+		enc.metaProto.BitsPerEncodedValue = int64(bits.Len(uint(len(dict))))
+	default:
+		// Default to delta encoding.
+		enc.metaProto.Encoding = encodingpb.EncodingType_DELTA
+		enc.metaProto.BitsPerEncodedValue = int64(bits.Len(uint(max - min)))
+	}
+
+	if err := proto.EncodeIntMeta(&enc.metaProto, &enc.buf, writer); err != nil {
+		return err
+	}
+
+	switch enc.metaProto.Encoding {
+	case encodingpb.EncodingType_DICTIONARY:
+		if err := enc.encodeDictionary(writer, enc.metaProto.BytesPerDictionaryValue, enc.metaProto.BitsPerEncodedValue, dict, valuesIt); err != nil {
+			return err
+		}
+	case encodingpb.EncodingType_DELTA:
+		if err := enc.encodeDelta(writer, enc.metaProto.BitsPerEncodedValue, valuesIt); err != nil {
+			return err
+		}
+	}
+
+	return nil
 }
 
 // Reset resets the encoder.
-func (enc *IntEnc) Reset() { panic("not implemented") }
+func (enc *IntEnc) Reset() {
+	enc.metaProto.Reset()
+	enc.dictionaryProto.Reset()
+}
+
+func (enc *IntEnc) encodeDelta(
+	writer io.Writer,
+	bitsPerEncodedValue int64,
+	valuesIt RewindableIntIterator,
+) error {
+	// Get first record and set as the start.
+	valuesIt.Next()
+	curr := valuesIt.Current()
+	enc.metaProto.DeltaStart = int64(curr)
+
+	// Encode the first value which is always a delta of 0.
+	bitWriter := bitstream.NewWriter(writer)
+	// Write an extra bit to encode the sign of the delta.
+	if err := bitWriter.WriteBits(uint64(0), int(bitsPerEncodedValue)+1); err != nil {
+		return err
+	}
+
+	var delta int
+	last := curr
+	for valuesIt.Next() {
+		curr = valuesIt.Current()
+		delta = curr - last
+		if delta < 0 {
+			// Flip the sign.
+			delta *= -1
```

Comment:
`delta = -delta` is probably more efficient here.

Reply:
👍